[Dynamo] Trace autograd.function in dynamo when inputs require grad (#116358)

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116897

For training graphs (when inputs require grad), previously, we would speculate the forward and backward graph to determine if there are any graph breaks, side effect and etc but would not actually use these speculated graphs. We would just insert a call function node on the graph and later rely on autograd's tracing.

This approach does not work for more generalized graphs like graphs that include user defined triton kernels because autograd is not able to do the higher order function conversation.

This PR speculates the forward and backward functions and emits them in a HOF that later gets used via templating mechanism.

While working on this PR, I have exposed some bugs in the current tracing due to trampoline functions losing the source information resulting in incorrect graphs being produced. I have fixed these source information bugs and killed the trampolines.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

oulgen committed:
[Dynamo] Trace autograd.function in dynamo when inputs require grad (#116358)

For training graphs (when inputs require grad), previously, we would speculate the forward and backward graph to determine if there are any graph breaks, side effect and etc but would not actually use these speculated graphs. We would just insert a call function node on the graph and later rely on autograd's tracing.

This approach does not work for more generalized graphs like graphs that include user defined triton kernels because autograd is not able to do the higher order function conversation.

This PR speculates the forward and backward functions and emits them in a HOF that later gets used via templating mechanism.

While working on this PR, I have exposed some bugs in the current tracing due to trampoline functions losing the source information resulting in incorrect graphs being produced. I have fixed these source information bugs and killed the trampolines.

[ghstack-poisoned]

oulgen commented:
@Skylion007 --  @jansel already approved this PR on (https://github.com/pytorch/pytorch/pull/116358) but some internal tests were failing, so it got reverted

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1531,3 +1426,206 @@ def call_function(
+                tracer=bwd_tracer,
+            )
+
+        # TODO: assert that bwd_graph didn't capture values that were
+        # not created inside fwd_graph.
```
We probably want to do this soon, otherwise it'll lead to silent incorrectness

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1531,3 +1426,206 @@ def call_function(
+            fwd_args,
+            kwargs,
+            "autograd.Function",
+            enable_grad=False,
```
Removing this line seems to get the internal test case to pass, but I'm not sure why, morally enable_grad=False is correct because that's what autograd.Function does...

oulgen committed:
Update on "[Dynamo] Trace autograd.function in dynamo when inputs require grad (#116358)"


For training graphs (when inputs require grad), previously, we would speculate the forward and backward graph to determine if there are any graph breaks, side effect and etc but would not actually use these speculated graphs. We would just insert a call function node on the graph and later rely on autograd's tracing.

This approach does not work for more generalized graphs like graphs that include user defined triton kernels because autograd is not able to do the higher order function conversation.

This PR speculates the forward and backward functions and emits them in a HOF that later gets used via templating mechanism.

While working on this PR, I have exposed some bugs in the current tracing due to trampoline functions losing the source information resulting in incorrect graphs being produced. I have fixed these source information bugs and killed the trampolines.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh commented:
Does this change mean that pre-dispatch graphs will not longer have to see a higher order `AutogradFunction` op in the graph anymore, since we're inlining the contents? (@zou3519)

zou3519 commented:
> Does this change mean that pre-dispatch graphs will not longer have to see a higher order `AutogradFunction` op in the graph anymore, since we're inlining the contents? (@zou3519)

Pre-dispatch graphs will still see a higher order `AutogradFunction` op.

oulgen committed:
Update on "[Dynamo] Trace autograd.function in dynamo when inputs require grad (#116358)"


For training graphs (when inputs require grad), previously, we would speculate the forward and backward graph to determine if there are any graph breaks, side effect and etc but would not actually use these speculated graphs. We would just insert a call function node on the graph and later rely on autograd's tracing.

This approach does not work for more generalized graphs like graphs that include user defined triton kernels because autograd is not able to do the higher order function conversation.

This PR speculates the forward and backward functions and emits them in a HOF that later gets used via templating mechanism.

While working on this PR, I have exposed some bugs in the current tracing due to trampoline functions losing the source information resulting in incorrect graphs being produced. I have fixed these source information bugs and killed the trampolines.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
Import to fbsource or ghimport did not work but i imported internally using wget patch and apply: D52706773
All internal tests passing now.

oulgen commented:
@pytorchbot merge
